compiler/rustc_target/src/callconv/x86_win64.rs

```diff
@@ -41,10 +42,12 @@ pub(crate) fn compute_abi_info<Ty>(cx: &impl HasTargetSpec, fn_abi: &mut FnAbi<'
     }
     for arg in fn_abi.args.iter_mut() {
         if arg.is_ignore() {
-            // x86_64-pc-windows-gnu doesn't ignore ZSTs.
+            // x86_64-pc-windows-gnu doesn't ignore ZSTs -- except in `extern "win64"`
+            // functions which are supposed to match MSVC.
             if cx.target_spec().os == "windows"
                 && cx.target_spec().env == "gnu"
                 && arg.layout.is_zst()
+                && fn_abi.conv != Conv::X86_64Win64
```

Comment:
this needs to also cover at least fastcall and vectorcall which have basically the same issue: https://clang.godbolt.org/z/MjjfPo3rx

Reply:
Using fastcall and vectorcall with GCC causes the binaries to segfault, so GCC might be wrong here...

Reply:
hmm, at least gcc matches clang for the short functions i tested above...

Reply:
Do we have anything matching `__attribute__((sysv_abi))`?

If the binaries segfault then maybe it's not so clear what fastcall and vectorcall should do?

Reply:
Strangely, the test fails for fastcall on windows-gnu (i.e., we do pass the ZST argument there). I don't yet understand why.

Reply:
Ah, we currently have explicit logic in `adjust_abi` which says that on x86-64, "fastcall" should be like "C". This is obviously in contradiction with saying that "fastcall" should ignore ZST while "C" should pass them by-pointer.

I'm not a Windows expert so I don't know what is correct here.

Reply:
> this needs to also cover at least fastcall and vectorcall which have basically the same issue: https://clang.godbolt.org/z/75fo4cbhn

Odd, so vectorcall was correct before this PR already? Not quite sure how that is happening.

Reply:
Anyway the latest version should handle vectorcall and fastcall, hopefully.

Reply:
> perhaps, however I think it's at least worth matching clang where they do work.

Sure, just pointed out this is a mess with GCC.
A couple years ago I noticed some fastcall tests disabled on windows-gnu because they were failing when objects were built with GCC but worked fine when built with Clang. So yeah, while GCC de facto defines the standard, it's broken in many ways.